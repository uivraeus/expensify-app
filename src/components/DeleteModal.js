import React from 'react';
import Modal from 'react-modal';

const DeleteModal = (props) => (
  <Modal 
    isOpen={props.queryConfirm}
    onRequestClose={props.onNo}
    contentLabel="Confirm deletion"
    closeTimeoutMS={200}
    className="modal"
  >
    <h3 className="modal__title">Do you really want to delete the expense?</h3>
    <div className="modal__button-group">
      <button className="button" onClick={props.onNo}>No</button>
      <button className="button" onClick={props.onYes}>Yes</button>
    </div>
  </Modal>
);

export default DeleteModal;