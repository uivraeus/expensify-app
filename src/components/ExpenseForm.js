import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
// assume that a parent has imported css for date-picker (see app.js)

//const now = moment();
//console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount/100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      errorMessage: ''
    };
  }
 

  // NOTE! store e.target.value in local variable as it can not be referenced 
  // in the setState callback
  // (possible to solve by setting the event to "persist" via e.persist() )
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState( () => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState ( () => ({ note }));
  };

  onAmountChange = (e) => {
    // format validation (empty string OR only number with max 2 (optional) decimals)
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState( () => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    // actuallyu not needed for my (for some reason), but...
    // if field was cleared (DEL-key) then don't update
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({focused}) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault(); //prevent page refresh

    //form validation (non-empty description and amount)
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({errorMessage: 'Please provide description and amount'}));
    } else {
      this.setState(() => ({errorMessage: ''}));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        <p>{this.state.errorMessage}</p>
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date = {this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false} //possible to select dates in the past
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>Submit</button>
        </form>
      </div>
    );
  }
}
