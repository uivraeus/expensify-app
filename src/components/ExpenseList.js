import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// "export" for testing 
export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div className="list-item list-item__message">
            <span>No Expenses</span>
          </div>
        ) : (
          props.expenses.map( (expense) => (
            <ExpenseListItem key={expense.id} {...expense}/>
          ))
        )
      }
    </div>
        
  </div>
);


//the "state" arg is the store's state, i.e. what one gets when calling store.getState()
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
}

//connect() returns a function which is used to create the HOC
export default connect(mapStateToProps)(ExpenseList);

