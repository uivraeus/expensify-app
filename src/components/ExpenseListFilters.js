import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

// Some refactoring applied to make testing easier, e.g. use class methods instead
// of inline functions for callbacks.
// Also use mapDispatchToProps (see below) instead of direct calls to actions from within 
// the class.
export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  onSortChange = (e) => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  }

  render() {
    return(
      <div>
        <input 
          type="text"
          value={this.props.filters.text} onChange={this.onTextChange}
        />
        <select
          value={this.props.filters.sortBy} onChange={this.onSortChange}
        >
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>  
    )
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);