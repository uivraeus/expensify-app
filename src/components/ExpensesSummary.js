import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

// Export for testing
export const ExpensesSummary = (props) => {
  const expenseStr = props.expenseCount === 1 ? 'expense' : 'expenses';
  const totalStr = numeral(props.expensesTotal/100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Viewing <span>{props.expenseCount}</span> {expenseStr} totalling <span>{totalStr}</span></h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

// the store's state -> props in ExpensesSummary
const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return ({
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  });
};

export default connect(mapStateToProps)(ExpensesSummary);