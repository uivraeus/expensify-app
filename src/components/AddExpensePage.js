import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

//NOTE:
//mapDispatchToProps used below to avoid referencing startAddExpense() [actions/]
//directly from the AddExpensePage. This makes it easier to test AddExpensePage as
//the "props.startAddExpense" function can be handled by a spy

//NOTE II:
//Why a class based component when no state is managed?
//- have onSubmit as a member method instead of (re-)defining it everytime
//  the component is rendered

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/'); //goto dashboard
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

//The logic in this map-function will not be covered by the test of AddExpensePage
//i.e. it is replaced by a spy
const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage); //get access to dispatch()
