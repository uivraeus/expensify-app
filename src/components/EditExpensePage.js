import React from 'react'; 
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import DeleteModal from './DeleteModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

// Refactored with mapDispatchToProps and class-based component as done
// for AddExpensePage (see that one for futher details)

export class EditExpensePage extends React.Component {
  state = {
    queryDelete: false
  };

  onClick = () => {
    this.setState(() => ({queryDelete: true}))  
  };

  onDeleteConfirm = () => {
    this.setState(() => ({queryDelete: false}))
    
    this.props.startRemoveExpense(this.props.expense.id)
    this.props.history.push('/'); //goto dashboard
  }

  onDeleteAbort = () => {
    this.setState(() => ({queryDelete: false}))
  }

  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/'); // /dashboard... or, route to / and rely on re-routing when loged in?
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button className="button button--secondary" onClick = {this.onClick}>Remove Expense</button>
        </div>
        <DeleteModal queryConfirm={this.state.queryDelete} onYes={this.onDeleteConfirm} onNo={this.onDeleteAbort}/>
      </div>
    );
  };
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) =>  expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);