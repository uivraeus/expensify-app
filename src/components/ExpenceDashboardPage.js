import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenceDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpensesSummary />
    <ExpenseList />
  </div>
);

export default ExpenceDashboardPage;