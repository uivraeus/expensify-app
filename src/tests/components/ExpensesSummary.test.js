import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render expenses summary (no expenses)', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={0} expensesTotal={0} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render 1 expense summary', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={12350} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render 2 expenses summary', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={2} expensesTotal={22350} />);
  expect(wrapper).toMatchSnapshot();
});