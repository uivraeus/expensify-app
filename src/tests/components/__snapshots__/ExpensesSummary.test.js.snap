// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1 expense summary 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $123.50
  </h1>
</div>
`;

exports[`should render 2 expenses summary 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totalling 
    $223.50
  </h1>
</div>
`;

exports[`should render expenses summary (no expenses) 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
     totalling 
    $0.00
  </h1>
</div>
`;
